### DISCUSSION TITLE
How to provide rules to the library?

### DISCUSSION THREAD

**Discussion by CurlyMoo** [CurlyMoo/rules#2] [Q&A]
At the moment, the rule parses just parses rule provided to it. The developer should handle the list of rules to be parsed or (re)executed. Or should the rule library have the possibility to read rules from external locations such as files?

@TD-er commented on this:

> I haven't looked at your code in great depth as of now, but I was thinking maybe I can add some wrapper to call the rules parsing with some range/location parameters.
> For example an offset in a file, or maybe you can add some jump condition based on a hash of some event name?

**Comment by TD-er** [comment]
Well it depends I guess.
Your rules currently have a different syntax to the ESPEasy rules.
So the first step would probably be to "convert" the ESPEasy rules into a syntax that's easier to parse.
I don't know where this responsibility should be.
If you intend to have this rules project independent of ESPEasy, then it makes sense "I" translate it and thus the rules parser does not read from the rules directly.

Apart from this, there is also the question on how to access the parameters that have to be replaced at runtime during rules parsing?
For example, does the rules handler have direct access to `UserVar` in ESPEasy? Or should there be a call for "give me `[taskname#varname]` value"?
The same applies for system variables, but those can be encoded using an enum I guess.

The current ESPEasy rules now support something like this `[var#%v1%]` so that's a 2-step replacement.

**Reply by CurlyMoo** [reply]
There are multiple discussions going on here:

1. What should the future ESPEasy syntax look like and is it discussable to change it? If not, how do we allow a generic rule library parse different dialects?
2. How can a rule library have the largest level of independence but still allow for implementation in various projects.
3. What feature should the library incoorporate. I would like to discuss this in a seperate discussion.

**Reply by TD-er** [reply]
About the syntax:
The syntax of rules in ESPEasy is being used by lots and lots of users and explained in numerous blog posts.  A good way to chase users away is by rendering all their effort void.  Thus if things need to be changed, I think we should 'translate' it before serving it to the rules handler/parser.
I recently learned Tasmota also has some rules handling which does resemble the ESPEasy syntax quite a bit.
So 2 of the biggest firmwares for ESPs do have a slightly different syntax from this rules parser.

N.B. this translation could be done in several ways.
The most flexible is embedded in the code of either ESPEasy or this rules parser thus all code that should handle the ESPEasy syntax should then perform the translation and hand it over to this rules parser.
Another way could be to allow a one time conversion, which can either be a special build of ESPEasy (intermediate build for conversion) or some JavaScript which can be executed in the browser to download the "old" rules and save it in the "new" format. But this has a huge drawback as it needs new documentation of all the rules we currently have.

Point 2:
A translation layer does make it the most flexible, but like I said the syntax used by Tasmota does resemble ESPEasy's rules syntax.
Maybe you should also think about other existing projects like graphical code editors (blockly???, not sure about the name), what kind of syntax can they generate.
Also Node-RED like projects are very popular. Not sure how they generate code.

**Reply by CurlyMoo** [reply]
Let's invite @arendst as well to see what he thinks about a new rule library, and what issues he might have had with the current one.

>  I recently learned Tasmota also has some rules handling which does resemble the ESPEasy syntax quite a bit.

True, they copied it from you guys.

All of the rest you're saying is true, but please consider that that syntax was though of from a (no-offense) pretty simple rule parsing perspective. I still love the fact that it exists, and is working well. But the current syntax gives a lot of headaches as well. You can see what i currently built for the Heatpump here: https://gathering.tweakers.net/forum/list_message/65401072#65401072

General theories about lexers, parsers, abstract syntax tree etc. just don't exist in the current code. And there is a reason why common programming languages in the core all have pretty similar syntax.  I don't know any language that allows you to call a function or either set a variable in the manner ESPEasy does: `taskValueSet 1,1,0`.

So, i would suggest that we either:
- Built a translater that translates the current syntax to the new syntax 
- Give the user the possibility to use the new syntax next to the old one, as you currently do.
- Switch to the new syntax.

>  I recently learned Tasmota also has some rules handling which does resemble the ESPEasy syntax quite a bit.

It would be great if we could stop implementing a slighty different rule system in each (embedded) platform, but just have one library that works well across all of them, and therefor also shares documentation. In that case, the documentation can just tell you what the small differences are between Tasmota, ESPEasy or ESPHome, and that's.

**Reply by CurlyMoo** [reply]
Even if we would opt for using the new syntax together with a translator, we still need to think the current syntax through properly. So we don't have the same issue in the future. That some developer pops in and tells us the current new rule system is bogus 😉

**Reply by TD-er** [reply]
> That some developer pops in and tells us the current new rule system is bogus

Well that's always going to happen :)
I don't think there is a more stubborn kind of person like a programmer :)

Not sure if Theo is even considering switching rules engine. I only mentioned it as a comparisson to "defend" the ESPEasy syntax even though I don't think it is a good syntax/grammar. It was only to point out there currently is quite a large user base using it right now.

**Reply by CurlyMoo** [reply]
> It was only to point out there currently is quite a large user base using it right now.

I understand. I will check if i can easily make a lexer for the ESPEasy syntax. Can you provide me with a good example rule to test with?

**Reply by AloisKlingler** [reply]
a big 👍 for the ruleset of the heatpump. and I thought my shuttercontrol-ruleset is kind of "do not do this with the rules". :-)
If you guys want the opinion from an end-user with scripting and slightly programming experience: what a pity, there is no Block-based coding GUI for the current rules, because it is exactly what I had to do by hand.
not because I am not able to write if's or "functions". because a lot people will be able to understand and use it as soon as the do not need any programming experience. with block-based coding the obstacle is very low.

"outsourcing" functions to the event-based approach looked for me like: "yes, we can handle it, less memory on the ESP or less developing time brings me in the situation to attach this simply in the already existing event-logic."

@TD-er the existing users will not go away just by switching the rules engine. create another build which has still the "old" rules engine in place, mark it as legacy with a link for new rules.
give a (web based?) possibility to transform existing rules (not on the ESP). there should not be much magic behind.
give a web based possibility to create block-based coding for new rules only (import by e.g. json the esp-config to have the devices and controllers as blocks available - oh, how cool would that be!). this will be the breaking point where users stop typing and start clicking. ;)

**Reply by CurlyMoo** [reply]
Just to give you a sneek peek. This is how i (still partly) moving my current heatpump ESPEasy code to the new rules library.

```
on setThermostaat then
	#nachtverlaging = -2;
	#T1 = 21.25;
	#T2 = 20.25;
end

on @Room_Thermostat_Temp then

	setThermostaat();

	$Taanvoer1 = 30;
	$Tbuiten1 = 12;
	$Taanvoer2 = 41;
	$Tbuiten2 = -6;

	$MaxTa = 0;

	$Temp = @Room_Thermostat_Temp;

	$targetTempDiff = #T1 - $Temp;

	if @Outside_Temp >= $Tbuiten1 then
		$MaxTa = $Taanvoer1;
	else
		if @Outside_Temp <= $Tbuiten2 then
			$MaxTa = $Taanvoer2;
		else
			$MaxTa = $Taanvoer1 + (($Tbuiten1 - @Outside_Temp) * ($Taanvoer2 - $Taanvoer1) / ($Tbuiten1 - $Tbuiten2));
		end
	end

	if $Temp > #T1 then
		@SetHeatPump = 0;
	else
		if $Temp < #T2 then
			@SetHeatPump = 1;
		end
	end

	if @Heatpump_State == 1 then
		if @Outside_Temp > 0 then
			if @Compressor_Freq == 0 && @Defrosting_State == 0 then
				@Main_Outlet_Temp = @Main_Outlet_Temp + 5;
			else
				@Main_Outlet_Temp = @Main_Outlet_Temp - 1;
			end
		end
	end
end
```

In this case:
`@` = heatpump command
`#` = global variable
`$` = local variable

**Comment by CurlyMoo** [comment]
The latest changes indeed allows the library to parse multiple rule blocks in a single *file*. This is intentionally not all done inside the rules library. This prevent a long blocking function call or the rule initialization.

**Comment by CurlyMoo** [comment]
My fully functional rules library inside an working program, which should replace the ESPEasy implementation can be found here:
https://github.com/CurlyMoo/HeishaDaemon

In the `rules.txt` you can see the actual new syntax fully implemented.

In the `main.cpp` source file you can see i how modularly dealt with `events`, `timers`, `globals`, `variables` and `system values`.

**Comment by CurlyMoo** [comment]
@TD-er The last version of my library is the best i could do after testing various implementations. E.g. what i tried with commits 4db5a4befd6156cb57e127f5f2c313176220fc87, 1137227d90dbc5eda94c78f1f66d0b1e3833d309, 1d3687ceaf0bdfc5515d92d419ec4621bbeaac01, 759f0350575ee606f36e56705add13c008ad43e7 which finally lead to commit 7863ea54e7739f62514819dfdb530a07bc23e195

I've been testing this last code on my ESP8266 and it already runs fine for hours with no exceptions.

The accompanied README describe pretty well how the whole thing works.

**Comment by TD-er** [comment]
I just browsed through the code a bit and I'm not really sure where to start to even get clear for myself what still has to be done to get this integrated in ESPEasy.

Things I can see so far:
- Transform rules syntax of ESPEasy into what is expected by this rules parser (e.g. adding `;`, marking commands to be like functions, if...then etc)
- Feed the rules into the rules parser to create interpreted byte code
- What to call to execute the rules given a specific event?
- How does the rules parser call commands, read values, etc?

**Reply by CurlyMoo** [reply]
> What to call to execute the rules given a specific event?

That's easily implemented like this (and how i implemented it already in my test setup):

https://gist.github.com/CurlyMoo/0ba1bceb26913266ba5c6c8df6c349f1#file-start-ino-L215
https://gist.github.com/CurlyMoo/0ba1bceb26913266ba5c6c8df6c349f1#file-start-ino-L126

> How does the rules parser call commands, read values, etc?

Read (global) values (`$Var`)
https://gist.github.com/CurlyMoo/0ba1bceb26913266ba5c6c8df6c349f1#file-start-ino-L362

Globals (`#Temp`)
https://gist.github.com/CurlyMoo/0ba1bceb26913266ba5c6c8df6c349f1#file-start-ino-L413

E.g. Heatpump commands (`@Heatpump_State`)
https://gist.github.com/CurlyMoo/0ba1bceb26913266ba5c6c8df6c349f1#file-start-ino-L482

System values (`%hour`)
https://gist.github.com/CurlyMoo/0ba1bceb26913266ba5c6c8df6c349f1#file-start-ino-L507

And the allowed variable syntax:
https://gist.github.com/CurlyMoo/0ba1bceb26913266ba5c6c8df6c349f1#file-start-ino-L135

Calling other code blocks:
https://gist.github.com/CurlyMoo/0ba1bceb26913266ba5c6c8df6c349f1#file-start-ino-L215

https://github.com/CurlyMoo/rules/blob/main/rules.cpp#L3628

https://github.com/CurlyMoo/rules/blob/main/rules.cpp#L4319

```
on foo then
  ...
end

on bar then
  foo();
end
```

----

Conclusion is indeed my documentation lacks the API explanation and syntax rules 👍

**Reply by CurlyMoo** [reply]
I've updated the documentation with explanation about event callbacks.

**Reply by CurlyMoo** [reply]
@TD-er I've put a great deal in improving the README.md. Please let me know what questions you still have or where you are stuck grasping the whole.

Next step i'm planning is to provide some really basic examples for your questions like how to deal with events and how to support global variables.

**Reply by TD-er** [reply]
Right now, I don't have time to dive into it, I hope to have time this weekend...

**Reply by CurlyMoo** [reply]
Take your time. In the meanwhile i've added two example to illustrate most of your questions.

**Reply by CurlyMoo** [reply]
That were some busy weekends :wink:

**Reply by TD-er** [reply]
Extremely busy and still is....

**Reply by CurlyMoo** [reply]
:pensive:
